File: 023-Merge-k-Sorted-Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        def merge(l1, l2):
            if l1 == l2: return l1
            else:
                dummy = ListNode(0)
                new_node = dummy
                while l1 != None and l2 != None:
                    if l1.val < l2.val:
                        new_val = l1.val
                        new_node.next = ListNode(new_val)
                        l1 = l1.next
                        new_node = new_node.next
                    else:
                        new_val = l2.val
                        new_node.next = ListNode(new_val)
                        l2 = l2.next
                        new_node = new_node.next

                if l1 == None:
                    new_node.next = l2
                else:
                    new_node.next = l1
                return dummy.next
        
        k = len(lists)
        if k == 0: return None
        elif k == 1: return lists[0]
        #elif k == 2: return merge(lists[0], lists[1])
        else:
            l1 = self.mergeKLists(lists[0:k//2])
            l2 = self.mergeKLists(lists[k//2:k])
            return merge(l1, l2)
            

File: 023-Merge-k-Sorted-Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size() == 0) return {};
        int amount = lists.size();
        int interval = 1;
        while(interval < amount)
        {
            for(int i = 0; i < amount - interval; i += (2 * interval))
            {
                lists[i] = mergeTwoLists(lists[i], lists[i + interval]);
            }
            
            interval *= 2;
        }
        
        return lists[0];
    }
    
    ListNode* mergeTwoLists(ListNode *&l1, ListNode *&l2)
    {
        ListNode *root = new ListNode(0);
        ListNode *p = root;
        while(l1 != NULL && l2 != NULL)
        {
            if(l1 -> val < l2 -> val)
            {
                p -> next = l1;
                l1 = l1 -> next;
            }
            
            else
            {
                p -> next = l2;
                l2 = l2 -> next;
            }
            
            p = p -> next;
        }
        
        if(l1 != NULL) p -> next = l1;
        else p -> next = l2;
        return root -> next;
    }
};

// Runtime: 24 ms, faster than 90.61% of C++ online submissions for Merge k Sorted Lists.
// Memory Usage: 11.2 MB, less than 63.09% of C++ online submissions for Merge k Sorted Lists.

// Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.

// Example:

// Input:
// [
//   1->4->5,
//   1->3->4,
//   2->6
// ]
// Output: 1->1->2->3->4->4->5->6
